blog/_posts/2018-04-12-in-a-nutshell.md

```diff
@@ -0,0 +1,93 @@
+---
+layout: blog-page
+title: Scala in a (tasty) Nutshell
+author: Martin Odersky and Nicolas Stucki
+authorImg: /images/martin.jpg
+date: 2018-03-05
+---
+
+One of the biggest open questions for migrating to Scala 3 is what to do about macros. In this blog post we give our current thinking, which is to try to achieve full alignment between macros and Tasty.
+
+## What is TASTY?
+
+Tasty is the high-level interchange format for Scala 3. It is based on <u>t</u>yped <u>a</u>bstract <u>s</u>yntax <u>t</u>rees. These trees contain in a sense all the information present in a Scala program. They represent the syntactic structure of programs in every detail and also contain the complete information about types and positions. The Tasty "snapshot" of a code file is taken after type checking (so that all types are present and all implicits are elaborated) but before any transformations (so that no information is lost or changed). The file representation of these trees is heavily optimized for compactness, which means that we can generate full Tasty trees on every compiler run and rely on nothing else for supporting separate compilation.
+
+The information present in TASTY trees can be used for many purposes.
+
+ - The compiler uses it to support separate compilation.
+ - A language server for an IDE uses it to support hyperlinking, command completion, or documentation.
+ - A build tool can use it to cross-build on different platforms and migrate code from one
+   version to another.
+ - Optimizers and analyzers can use it for deep code analysis and advanced code generation
+
+Of these usages, the first two work today. The other two are worthwhile possibilities to pursue in the future.
+
+OK, but what is Tasty _exactly_? An up-to-date version of the Tasty file format is described in file
+[TastyFormat.scala](https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/dotc/core/tasty/TastyFormat.scala)
+of the `dotc` compiler for Scala 3.
+
+## What Does it Have to Do with Macros?
+
+It turns out that Tasty also makes an excellent foundation for a new generation of reflection-based macros.
+
+The first problem with the current `scala.reflect` macros is that they that they are completely dependent on the current Scala compiler (internally named `nsc`). In fact, `scala.reflect` macros are nothing but a thin veneer on top of `nsc` internals. This makes them very powerful but also fragile and hard to use. Because of this, they have had "experimental" status for their whole lifetime. Since Scala 3 uses a different compiler (`dotc`), the old reflect-based macro system cannot be ported to it, so we need something different, and hopefully better.
+
```

Comment:
>  This makes them very powerful but also fragile and hard to use.

It would be good to have an analysis of _why it is hard to use_. I try to sketch some of the context. Maybe @xeno-by can add more and/or correct me.

For me, macros is a spectrum that can be at two extremes:

1. those that can be put on a solid foundation (e.g. based on PMP)
2. those that deal with low-level compiler entities: trees/types/symbols

It’s a big progress to identify a set of macros that can be implemented based on solid principles.

For those that cannot be based on solid principles, what do we do?

There are many macros that is out of the reach of principled handling, we cannot afford each one to be a compiler plugin. Also, portability seems still possible for some of the unprincipled macros. It seems we need a low-level portable macro system.

However, from the experience with Gestalt,  I'm pessimistic about a _generic portable macro system_, due to complex compiler invariants and usability concerns. A generic portable macro system forms an obstacle for communication and learning when the abstraction breaks. And the abstraction of a generic portable macro system is doomed to break due to the wide spectrum of macros.

Advanced macros are no different from compiler (or compiler plugin) development in nature. Thus, compiler crashes in developing such macros are expected. However, if there exists an abstraction layer in case of compiler crashes, compiler writers cannot help macro authors directly because of the existence of the abstraction layer. There are more compiler experts than macro system experts, it's a big disadvantage that they do not share the same body of knowledge. A generic portable macro system thus inevitably leads to user support problem. Programmers who write macros that break compiler invariants are usually advanced Scala programmers, it is not an issue for them to learn if there is enough guidance and support.

Instead, I'm optimistic for _domain-specific (semi-)portable macro systems_. Despite a generic abstraction is hopeless, the abstraction can be successful for a particular domain, like [magnolia](https://github.com/propensive/magnolia).  Domain-specific macro systems provide better programming experience because they can make more solid assumptions for a particular domain.

Then the question is, what is needed from the compiler side to support domain-specific (semi-)portable macro systems? One simple answer is that they are compiler plugins. But compiler plugins are a little heavier to set up and use, why not just expose compiler internals as a low-level _unportable_ macro system to support high-level domain-specific (semi-)portable macro systems? I see big advantages of doing so:

1. ease to implement in the compiler.
2. authors of advanced macros, compiler plugins and compiler hackers share the same body of knowledge.
3. makes development of high-level domain-specific (semi-)portable macro systems easier.

My vision for a future Scala meta-programming is the co-existence of (1) low-level unportable macro system; (2) several domain-specific macro systems (like magnolia) and (3) principled meta-programming systems (like inlining and PMP).

Reply:
Very well said! I always have a hard time remembering what things are called in scala.reflect land, because they are slightly different from the compiler internals. 

The other key point I agree with: we should not lose sight of identifying the real challenges faced by macro authors out there. For each of those, we should ask ourselves, does basing things on TASTY solve this? Don't get me wrong, I don't think basing things on TASTY is a bad idea, but it doesn't magically solve all our problems, so let's be clear about that. 

For example: hygiene; difficulties with getting owner & tree structure to align; embedding untyped trees into typed ones....  These are all difficult API design challenges. I'm sure there are more, and I'm not the one who has this list in my head, but we should not lose the knowledge we've built up with the old system, and make sure new proposals at least explain how they deal with these things. Not in this blog post, of course, but since we're on the topic I just wanted to say "+1" :-)

Reply:
I agree we should look at the difficulties of existing macros, but probably not in this blog post. The point is that the new approach is really a different world. For instance, regarding the obvious problems Adriaan mentioned:

- hygiene: not an issue in a typed world (I would not even know to define what it means!)
- owners should not be under programmer control, they should be handled fully automatically from the tree structure.
 - embedding untyped trees into typed ones: will not be possible.

I think it's futile to try to compare with existing macros now. We have to flesh out the system first to see where problems might lie. Also: I think it would be great if the new macro system was completely independent of the underlying compiler except for tree extractors. If that involves some duplication of code (e.g. subtyping implemented twice), that's a price worth to pay for it, IMO.